give more context to custom schemas

chenguo opened the pull request:
Kind of a corner case, I have a schema with a custom field that holds a recursive instance of that same schema, and I got blocked on not having access to `obj` in my custom `_jsonschema_type_mapping` call

chenguo committed:
give more context to custom schemas

chenguo commented:
@fuhrysteve 
just checking on the status of this repo, I see another PR was opened mid Jan and hasn't had movement. Is this project still actively maintained?

fuhrysteve commented:
@chenguo Thanks for the PR! Yes it this project is still very much actively maintained, though I don't always review PRs right away

fuhrysteve commented on marshmallow_jsonschema/base.py:
```diff
@@ -254,7 +254,7 @@ def _get_python_type(self, field):
         """Get schema and validators for field."""
         if hasattr(field, "_jsonschema_type_mapping"):
-            schema = field._jsonschema_type_mapping()
+            schema = field._jsonschema_type_mapping(self, obj)
```
Hmm, the problem with this is that it would break backward compatibility with custom type support as currently documented:

https://github.com/fuhrysteve/marshmallow-jsonschema#custom-type-support

Is there a way we can achieve what your trying to do without breaking backward compatibility?

chenguo replied:
hmm yeah, i think a part of the difficulty here is actually the kind of ad-hoc nature of `jsonschema_type_mapping`, where there's no explicit definition of an interface so we can't just update some base definition like 

```
def _jsonschema_type_mapping(self, json_schema = None, obj = None):
  ...
```
to make this easily work for everyone.

I think similar to how you're using reflection to check if that `_jsonschema_type_mapping` even exists, I can use `inspect` to count the number of arguments and only pass the context in if the argument count supports it. This would preserve existing behavior but does make this interface a bit more loosey than it already is.

I think I'm ok with that. You could argue a proper solution requires a revamp of how this functionality is supported anyhow, so the best we can do atm is just to bolt more on to what was bolted on.

I'll make an update, and also be sure to update the README this time.

chenguo committed:
use reflection to identify when new context functionlity is used

chenguo committed:
update readme

chenguo commented:
Ok, updated to not break existing interfaces, and added a README section to detail when this particular corner case is hit. Thanks for the feedback, lmk what else you need.

fuhrysteve commented:
Thanks @chenguo — picked up as commit `96bc062` in #200, using signature introspection to pass through the JSONSchema instance + obj only when the method declares 2 extra params. Closing this older PR.
